### PR TITLE
P2530R3 Hazard Pointers for C++26

### LINKED ISSUE #6293: [2023-06 LWG Motion 7] P2530R3 Hazard Pointers for C++26
https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p2530r3.pdf

### PR DESCRIPTION
**Pull request by jensmaurer**
Fixes #6293
Fixes cplusplus/papers#1194

~NOTE: This is based on LWG motion 6 and needs a rebase onto "main" during merging.~

### CODE REVIEW

File: source/threads.tex [hunk 1]
```diff
@@ -12006,3 +12006,455 @@
...
+Only the owner of the hazard pointer can set its value,
+while any number of threads may read its value.
+The owner thread sets the value of a hazard pointer to point to an object
+in order to indicate to concurrent threads---which
+may delete such an object---that
+the object is not yet safe to delete.
+
+\pnum
+A class type \tcode{T} is \defn{hazard-protectable}
+if it has exactly one base class of type \tcode{hazard_pointer_obj_base<T,D>}
```
Comment (JohelEGP) [inline source/threads.tex:12026]:
```suggestion
if it has exactly one base class of type \tcode{hazard_pointer_obj_base<T, D>}
```

File: source/threads.tex [hunk 2]
```diff
@@ -12006,3 +12006,455 @@
...
+for some \tcode{D},
+that base is public and non-virtual, and
+it has no base classes of type \tcode{hazard_pointer_obj_base<T2,D2>}
```
Comment (JohelEGP) [inline source/threads.tex:12029]:
```suggestion
it has no base classes of type \tcode{hazard_pointer_obj_base<T2, D2>}
```

File: source/threads.tex [hunk 3]
```diff
@@ -12006,3 +12006,455 @@
...
+for any other combination \tcode{T2}, \tcode{D2}.
+An object is \defn{hazard-protectable} if it is of hazard-protectable type.
+
+\pnum
+The time span between creation and destruction of a hazard pointer $h$
+is partitioned into a series of \defnadj{protection}{epochs};
```
Comment (JohelEGP) [inline source/threads.tex:12035]:
```suggestion
is partitioned into a series of \defnadjx{protection}{epochs}{epoch};
```

File: source/threads.tex [hunk 4]
```diff
@@ -12006,3 +12006,455 @@
...
+in each protection epoch,
+$h$ either is associated with a hazard-protectable object, or is unassociated.
```
Comment (JohelEGP) [inline source/threads.tex:12037]: The paper styles "associated with" and "unassociated" in italics.

File: source/threads.tex [hunk 5]
```diff
@@ -12006,3 +12006,455 @@
...
+Upon creation, a hazard pointer is unassociated.
+Changing the association (possibly to the same object)
+initiates a new protection epoch and ends the preceding one.
+
+\pnum
+An object \tcode{x} of hazard-protectable type \tcode{T} is
+\defn{retired} with a deleter of type \tcode{D}
+when the member function \tcode{hazard_pointer_obj_base<T,D>::retire}
```
Comment (JohelEGP) [inline source/threads.tex:12045]:
```suggestion
when the member function \tcode{hazard_pointer_obj_base<T, D>::retire}
```

File: source/threads.tex [hunk 6]
```diff
@@ -12006,3 +12006,455 @@
...
+The number of possibly-reclaimable objects has an unspecified bound.
+\begin{note}
+The bound can be a function of the number of hazard pointers,
+the number of threads that retire objects, and
+the number of threads that use hazard pointers.
+\end{note}
+\begin{example}
+The following example shows how hazard pointers allow updates to be carried out
+in the presence of concurrent readers.
+The object of type hazard_pointer in \tcode{print_name}
```
Comment (JohelEGP) [inline source/threads.tex:12101]:
```suggestion
The object of type \tcode{hazard_pointer} in \tcode{print_name}
```

File: source/threads.tex [hunk 7]
```diff
@@ -12006,3 +12006,455 @@
...
+void update_name(Name* new_name) {
+  Name* ptr = name.exchange(new_name);
+  ptr->retire();
+}
+\end{codeblock}
+\end{example}
+
+\rSec3[hazard.pointer.syn]{Header \tcode{<hazard_pointer>} synopsis}
+
+\begin{codeblock}
```
Comment (JohelEGP) [inline source/threads.tex:12124]:
```suggestion
\indexheader{hazard_pointer}%
\begin{codeblock}
```

File: source/threads.tex [hunk 8]
```diff
@@ -12006,3 +12006,455 @@
...
+namespace std {
+  // \ref{saferecl.hp.base}, class template \tcode{hazard_pointer_obj_base}
+  template<class T, class D = default_delete<T>> class hazard_pointer_obj_base;
+
+  // \ref{saferecl.hp.holder.general}, class \tcode{hazard_pointer}
```
Comment (JohelEGP) [inline source/threads.tex:12129]:
```suggestion
  // \ref{saferecl.hp.holder}, class \tcode{hazard_pointer}
```

File: source/threads.tex [hunk 9]
```diff
@@ -12006,3 +12006,455 @@
...
+\end{note}
+
+\pnum
+\complexity
+Constant.
+\end{itemdescr}
+
+\rSec4[saferecl.hp.holder.nonmem]{Non-member functions}
+
+\indexlibrarymember{swap}{hazard_pointer}%
```
Comment (JohelEGP) [inline source/threads.tex:12431]:
```suggestion
\indexlibraryglobal{make_hazard_pointer}%
```

**Review by JohelEGP — CHANGES_REQUESTED** [review]

File: source/support.tex [hunk 10]
```diff
@@ -635,6 +635,7 @@
   // also in \libheader{unordered_map}, \libheader{unordered_set}
 #define @\defnlibxname{cpp_lib_hardware_interference_size}@        201703L // also in \libheader{new}
 #define @\defnlibxname{cpp_lib_has_unique_object_representations}@ 201606L // also in \libheader{type_traits}
+#define @\defnlibxname{hazard_pointer}@                            202306L // also in \libheader{hazard_pointer}
```
Comment (JohelEGP) [inline source/support.tex:638]: I don't see this proposed in the paper.
```suggestion
#define @\defnlibxname{cpp_lib_hazard_pointer}@                    202306L // also in \libheader{hazard_pointer}
```
```suggestion
```

**Review by JohelEGP — CHANGES_REQUESTED** [review]

File: source/support.tex (hunk 10 above)
Comment (JohelEGP) [reply source/support.tex:638]: Maybe it's because you can just detect the presence of the header `<hazard_pointer>`.

Comment (JohelEGP) [reply source/support.tex:638]: Although that's not necessarily reliable,
as implementations provide empty headers when compiling with earlier standards.
I read something about that recently, but I can't find the source.

Comment (JohelEGP) [reply source/support.tex:638]: It turns out, like the other RCU paper,
that the proposed feature-test macro is very far from the wording.
§1.1 vs. §4.

File: source/threads.tex (hunk 1 above)
Comment (tkoeppe) [reply source/threads.tex]: Done.

File: source/threads.tex (hunk 5 above)
Comment (tkoeppe) [reply source/threads.tex]: Done.

File: source/threads.tex (hunk 3 above)
Comment (tkoeppe) [reply source/threads.tex:12051]: Done.

File: source/threads.tex (hunk 4 above)
Comment (tkoeppe) [reply source/threads.tex:12053]: That's true, but we don't always italicize local terms, e.g. see shared_ptr's "empty". I think this is fine.

Comment (tkoeppe) [reply source/threads.tex:12053]: On the other hand, "owns" below is a definition, so we might as well make this one one, too.

**Pull request merged**